lib/executor/variable_looping_vus.go

```diff
@@ -187,102 +182,87 @@ func (vlvc VariableLoopingVUsConfig) Validate() []error {
+	next := func(sign int64) int64 {
+		r := offsets[int(localIndex)%len(offsets)]
+		localIndex += sign
+		return r
+	}
+	i := start + 1 // this is the index for the full execution segment
 	for _, stage := range vlvc.Stages {
 		stageEndVUs := stage.Target.Int64
 		stageDuration := time.Duration(stage.Duration.Duration)
-		totalDuration += stageDuration
-
-		stageVUAbsDiff := abs(stageEndVUs - fromVUs)
-		if stageVUAbsDiff == 0 {
-			// We don't have to do anything but update the time offset
-			// if the number of VUs wasn't changed in this stage
-			timeFromStart += stageDuration
-			continue
-		}
-
-		// Handle 0-duration stages, i.e. instant VU jumps
-		if stageDuration == 0 {
-			fromVUs = stageEndVUs
-			prevScaledVUs = es.Scale(stageEndVUs)
-			steps = append(steps, lib.ExecutionStep{
-				TimeOffset: timeFromStart,
-				PlannedVUs: uint64(prevScaledVUs),
+		timeTillEnd += stageDuration
+
+		stageVUDiff := stageEndVUs - fromVUs
+		switch {
+		case stageDuration == 0:
+			addStep(lib.ExecutionStep{
+				TimeOffset: timeTillEnd,
+				PlannedVUs: uint64(et.ScaleInt64(stageEndVUs)),
 			})
-			continue
-		}
-
-		// For each stage, limit any VU adjustments between the previous
-		// number of VUs and the stage's target to happen at most once
-		// every minIntervalBetweenVUAdjustments. No floats or ratios,
-		// since nanoseconds should be good enough for anyone... :)
-		stepInterval := stageDuration / time.Duration(stageVUAbsDiff)
-		if stepInterval < minIntervalBetweenVUAdjustments {
-			stepInterval = minIntervalBetweenVUAdjustments
-		}
-
-		// Loop through the potential steps, adding an item to the
-		// result only when there's a change in the number of VUs.
-		//
-		// IMPORTANT: we have to be very careful of rounding errors,
-		// both from the step duration and from the VUs. It's especially
-		// important that the scaling via the execution segment should
-		// happen AFTER the rest of the calculations have been done and
-		// we've rounded the global "global" number of VUs.
-		for t := stepInterval; ; t += stepInterval { // Skip the first step, since we've already added that
-			if time.Duration(abs(int64(stageDuration-t))) < minIntervalBetweenVUAdjustments {
-				// Skip the last step of the stage, add it below to correct any minor clock skew
-				break
-			}
-			stepGlobalVUs := fromVUs + int64(
-				math.Round((float64(t)*float64(stageEndVUs-fromVUs))/float64(stageDuration)),
-			)
-			stepScaledVus := es.Scale(stepGlobalVUs)
-
-			if stepScaledVus == prevScaledVUs {
-				// only add steps when there's a change in the number of VUs
-				continue
+		case stageVUDiff != 0:
+			// Get the index to the start if they are not there
+			if i > fromVUs {
+				for ; i > fromVUs; i -= next(-1) {
+					if localIndex == 0 { // we want ot enter for this index but not actually go below 0
+						break
+					}
+				}
+			} else {
+				for ; i < fromVUs; i += next(1) { // <= test
+				}
```

Comment:
and we will just need to do in the `stageDuration == 0` case ... then ... or do you propose I run the code below and just have a logic that if I need to go through 20 steps there I just skip adding all but the last?

Reply:
Not sure... but the current code is very difficult to follow because it doesn't fit my mental model very well. Maybe the issue lies with me, but I'm fairly sure I understand what's supposed to happen here, and I find the current code very difficult to understand.

I imagine the `stages` as a series of staircases up and down, with maybe some flat surfaces (start and end VUs the same) or huge jumps (0 duration) from time to time. The `getRawExecutionSteps()` just emits the result of an imaginary person walking through that series of staircases... If there's no execution segment, we step on every stair. And if we have an execution segment, we just skip some of the stairs, according to the striping pattern.

So, I don't know what the cleanest way to implement this algorithm is, but this code doesn't seem to be it... One suggestion that comes to mind is that you might want to implement a local lambda version of `scaleInt64()` that doesn't throw away the value of `i`: https://github.com/loadimpact/k6/blob/762a2f0b4cae5cb52c69e9967efe5dc1cbb208ad/lib/execution_segment.go#L598-L604

This re-calculation is roughly what you're doing here, right? But you're basically doing it twice, once in the `et.ScaleInt64(stageEndVUs)` call above, and then the re-alignment here?